File: src/backend/app/schemas/kb_chat.py
"""KB chat schemas for iteration 129."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.schemas.rag import RAGDiagnostics
from app.schemas.stock_analysis import (
    StockAnalysisParams,
    StockAnalysisReportCard,
    StockAnalysisTaskCard,
)
from app.schemas.strategy import AIStrategyDraft

KBAssistantMode = Literal[
    "knowledge_qa",
    "strategy_idea",
    "backtrader_strategy",
    "strategy_review",
    "trading_execution",
    "stock_analysis",
]


class ConversationCreate(BaseModel):
    """Create conversation request."""

    knowledge_base_id: str = Field(..., min_length=1)
    title: str = Field("新对话", min_length=1, max_length=200)
    model_id: str | None = None

    @field_validator("title")
    @classmethod
    def normalize_title(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("title must not be blank")
        return normalized


class ConversationResponse(BaseModel):
    """Conversation response."""

    id: str
    knowledge_base_id: str | None = None
    user_id: str
    title: str
    model_id: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ConversationListResponse(BaseModel):
    """Conversation list response."""

    total: int
    items: list[ConversationResponse]


class ChatMessageResponse(BaseModel):
    """Chat message response."""

    id: str
    conversation_id: str
    role: str
    content: str
    citations: list[dict] | None = None
    tokens_used: int | None = None
    model_id: str | None = None
    reasoning: str | None = None
    metadata: dict[str, Any] | None = None
    assistant_mode: KBAssistantMode | None = None
    strategy_draft: AIStrategyDraft | None = None
    stock_analysis_task: StockAnalysisTaskCard | None = None
    stock_analysis_report: StockAnalysisReportCard | None = None
    reason_code: str | None = None
    diagnostic_message: str | None = None
    diagnostics: RAGDiagnostics | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

    @model_validator(mode="before")
    @classmethod
    def extract_metadata_fields(cls, value):
        if isinstance(value, dict):
            metadata = value.get("metadata") or value.get("metadata_json") or {}
            if isinstance(metadata, dict):
                value.setdefault("metadata", metadata)
                value.setdefault("assistant_mode", metadata.get("assistant_mode"))
                value.setdefault("strategy_draft", metadata.get("strategy_draft"))
                value.setdefault("stock_analysis_task", metadata.get("stock_analysis_task"))
                value.setdefault("stock_analysis_report", metadata.get("stock_analysis_report"))
                value.setdefault("reason_code", metadata.get("reason_code"))
                value.setdefault("diagnostic_message", metadata.get("diagnostic_message"))
                value.setdefault("diagnostics", metadata.get("diagnostics"))
            return value
        metadata = getattr(value, "metadata_json", None) or {}
        if not isinstance(metadata, dict):
            metadata = {}
        return {
            "id": getattr(value, "id", None),
            "conversation_id": getattr(value, "conversation_id", None),
            "role": getattr(value, "role", None),
            "content": getattr(value, "content", None),
            "citations": getattr(value, "citations", None),
            "tokens_used": getattr(value, "tokens_used", None),
            "model_id": getattr(value, "model_id", None),
            "reasoning": getattr(value, "reasoning", None),
            "metadata": metadata,
            "assistant_mode": metadata.get("assistant_mode"),
            "strategy_draft": metadata.get("strategy_draft"),
            "stock_analysis_task": metadata.get("stock_analysis_task"),
            "stock_analysis_report": metadata.get("stock_analysis_report"),
            "reason_code": metadata.get("reason_code"),
            "diagnostic_message": metadata.get("diagnostic_message"),
            "diagnostics": metadata.get("diagnostics"),
            "created_at": getattr(value, "created_at", None),
        }


class ChatHistoryResponse(BaseModel):
    """Conversation history response."""

    conversation_id: str
    messages: list[ChatMessageResponse]


class KBChatRequest(BaseModel):
    """Send chat message request."""

    knowledge_base_id: str | None = Field(None, min_length=1)
    question: str = Field(..., min_length=1, max_length=2000)
    conversation_id: str | None = Field(None, min_length=1)
    model_id: str | None = None
    assistant_mode: KBAssistantMode = "knowledge_qa"
    thinking_mode: bool = False
    stock_analysis_params: StockAnalysisParams | None = None

    @field_validator("question")
    @classmethod
    def normalize_question(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("question must not be blank")
        return normalized

    @model_validator(mode="after")
    def require_knowledge_base_for_qa(self):
        if self.assistant_mode == "knowledge_qa" and not self.knowledge_base_id:
            raise ValueError("knowledge_base_id is required for knowledge_qa")
        return self


class KBChatResponse(BaseModel):
    """Send chat response."""

    conversation_id: str
    answer: str
    citations: list[dict]
    context_chunks_used: int
    tokens_used: int
    model_id: str | None = None
    assistant_mode: KBAssistantMode = "knowledge_qa"
    strategy_draft: AIStrategyDraft | None = None
    stock_analysis_task: StockAnalysisTaskCard | None = None
    stock_analysis_report: StockAnalysisReportCard | None = None
    reasoning: str | None = None
    reason_code: str | None = None
    diagnostic_message: str | None = None
    diagnostics: RAGDiagnostics | None = None

File: src/frontend/src/api/kbChat.ts
import api from './index'
import type { ParamSpec } from '@/types'

export const KB_CHAT_SEND_TIMEOUT_MS = 120000

export type KBAssistantMode =
  | 'knowledge_qa'
  | 'strategy_idea'
  | 'backtrader_strategy'
  | 'strategy_review'
  | 'trading_execution'
  | 'stock_analysis'

export type KBReasonCode =
  | 'no_context_found'
  | 'ai_not_configured'
  | 'ai_provider_failed'
  | string

export interface KBStrategyDraftDataSource {
  type: string
  symbol?: string | null
  symbol_name?: string | null
  timeframe: string
  timeframe_n: number
  start_date?: string | null
  end_date?: string | null
  adjustment?: string | null
}

export interface KBStrategyDraftBacktestDefaults {
  initial_cash: number
  commission: number
  annual_days: number
  calc_method: string
  weight_mode: string
}

export interface KBStrategyDraftExecutionPlan {
  workspace_type: string
  group_name?: string | null
  run_parallel: boolean
}

export interface KBStrategyDraft {
  name: string
  description: string
  code: string
  params: Record<string, ParamSpec>
  category: string
  assumptions: string[]
  risk_points: string[]
  data_source: KBStrategyDraftDataSource
  backtest_defaults: KBStrategyDraftBacktestDefaults
  execution_plan: KBStrategyDraftExecutionPlan
  rationale?: string | null
  next_steps?: string[]
  suggested_symbol?: string | null
  suggested_timeframe?: string | null
}

export interface KBConversation {
  id: string
  knowledge_base_id?: string | null
  title: string
  model_id?: string | null
  created_at: string
  updated_at: string
}

export interface KBStockAnalysisParams {
  symbol: string
  market_type?: string
  analysis_date?: string
  research_depth?: string
  selected_modules?: string[]
  include_sentiment?: boolean
  include_risk?: boolean
  language?: string
  model_id?: string
}

export interface KBStockAnalysisTask {
  task_id: string
  symbol: string
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  progress: number
  current_step?: string | null
  message?: string | null
}

export interface KBStockAnalysisReport {
  report_id: string
  symbol: string
  summary: string
  decision_label: string
  risk_level: string
  confidence_score?: number | null
  export_formats: Array<'markdown' | 'html' | 'docx' | 'pdf'>
}

export interface KBConversationListResponse {
  total: number
  items: KBConversation[]
}

export interface KBCitation {
  document_id?: string | null
  document_title?: string | null
  chunk_id?: string | null
  chunk_index?: number | null
  similarity?: number | null
  content?: string | null
  score_breakdown?: Record<string, number> | null
}

export interface KBRetrievalDiagnostics {
  retrieval_profile: string
  search_mode: string
  search_query: string
  query_rewritten?: boolean
  applied_top_k: number
  applied_min_similarity: number
  history_messages_used?: number
  total_indexable_documents?: number
  indexed_documents?: number
  coverage_ratio?: number
}

export interface KBHistoryMessage {
  id: string
  conversation_id: string
  role: 'user' | 'assistant'
  content: string
  citations?: KBCitation[] | null
  tokens_used?: number | null
  model_id?: string | null
  assistant_mode?: KBAssistantMode | null
  strategy_draft?: KBStrategyDraft | null
  reasoning?: string | null
  reason_code?: KBReasonCode | null
  diagnostic_message?: string | null
  diagnostics?: KBRetrievalDiagnostics | null
  stock_analysis_task?: KBStockAnalysisTask | null
  stock_analysis_report?: KBStockAnalysisReport | null
  created_at: string
}

export interface KBHistoryResponse {
  conversation_id: string
  messages: KBHistoryMessage[]
}

export interface KBAskResponse {
  conversation_id: string
  answer: string
  citations: KBCitation[]
  context_chunks_used: number
  tokens_used: number
  model_id?: string | null
  assistant_mode?: KBAssistantMode
  strategy_draft?: KBStrategyDraft | null
  stock_analysis_task?: KBStockAnalysisTask | null
  stock_analysis_report?: KBStockAnalysisReport | null
  reasoning?: string | null
  reason_code?: KBReasonCode | null
  diagnostic_message?: string | null
  diagnostics?: KBRetrievalDiagnostics | null
}

export const kbChatApi = {
  listConversations(knowledgeBaseId?: string | null, params?: { skip?: number; limit?: number }) {
    const requestParams: Record<string, unknown> = { ...params }
    if (knowledgeBaseId) {
      requestParams.knowledge_base_id = knowledgeBaseId
    }
    return api.get<KBConversationListResponse>('/kb-chat/conversations', {
      params: requestParams,
    })
  },
  getHistory(conversationId: string) {
    return api.get<KBHistoryResponse>(`/kb-chat/history/${conversationId}`)
  },
  send(data: {
    knowledge_base_id?: string | null
    question: string
    conversation_id?: string | null
    model_id?: string
    assistant_mode?: KBAssistantMode
    thinking_mode?: boolean
    stock_analysis_params?: KBStockAnalysisParams
  }) {
    const payload: Record<string, unknown> = {
      question: data.question,
      conversation_id: data.conversation_id,
      model_id: data.model_id,
      assistant_mode: data.assistant_mode,
      thinking_mode: data.thinking_mode,
    }
    if (data.knowledge_base_id) {
      payload.knowledge_base_id = data.knowledge_base_id
    }
    if (data.stock_analysis_params) {
      payload.stock_analysis_params = data.stock_analysis_params
    }
    return api.post<KBAskResponse>('/kb-chat/send', payload, {
      timeout: KB_CHAT_SEND_TIMEOUT_MS,
    })
  },
}
